### PR TITLE
Multiple improvements to distributed code

### PR DESCRIPTION
**Pull request by ddkalamk**
- Align distributed embedding table initialization with single rank initialization
- Allow selecting alltoall impl using env DLRM_ALLTOALL_IMPL
- Skip batch when minibatch is not divisible by world size to avoid runtime error
- Add support for using Intel oneCCL custom backend
- Add support to specify output directory for profiler log files
- Override print function to print from rank 0 and provide print_all function to print from all the ranks
- Fix DDP wrapper when using GPUs
- Improve alltoall support check and non-mpi backend init

### CODE REVIEW

File: dlrm_s_pytorch.py [hunk 1]
```diff
@@ -142,7 +142,13 @@ def create_mlp(self, ln, sigmoid_layer):
 
     def create_emb(self, m, ln):
         emb_l = nn.ModuleList()
+        # save the numpy random state
+        np_rand_state = np.random.get_state()
         for i in range(0, ln.size):
+            if ext_dist.my_size > 1:
+                if not i in self.local_emb_indices: continue
+            # Use per table random seed for Embedding initialization
+            np.random.seed(self.l_emb_seeds[i])
```
Comment (dmudiger) [inline dlrm_s_pytorch.py:151]: We should probably also do this in the general case as well to keep things consistent (cc: @mnaumovfb) ?

**Comment by dmudiger** [comment]
LGTM, thanks for adding these changes @ddkalamk !

**Comment by ddkalamk** [comment]
Thanks @dmudiger !

File: dlrm_s_pytorch.py (hunk 1 above)
Comment (mnaumovfb) [reply dlrm_s_pytorch.py:151]: Can you add  a "WARNING: ..." comment explaining that the random seed is initialized per table in parallel mode (with MPI backend), while they are initialized once in the sequential run? I think for now this will be sufficient.

Comment (ddkalamk) [reply dlrm_s_pytorch.py:151]: The initialization (and per table random seeds) is consistent across sequential and parallel mode within this dist_exp branch. It only differs compared to master branch. So, this warning can be misleading...

Comment (mnaumovfb) [reply dlrm_s_pytorch.py:151]: Ok, understood. Thank you for clarifying. In that case let's leave it as is for now.
